R/cross.R

```diff
@@ -0,0 +1,160 @@
+#'
+#' When the number of combinations is large and the individual elements
+#' are heavy memory-wise, it is often useful to filter unwanted
+#' combinations on the fly. \code{.filter} must be a predicate
+#' functio! that takes the same number of arguments as the number of
+#' crossed objects (2 for \code{cross()}, 3 for \code{cross3()},
+#' \code{length(.l)} for \code{cross_n()}) and returns \code{TRUE} or
+#' \code{FALSE}. The combinations where the predicate function returns
+#' \code{TRUE} will be removed from the result.
+#'
+#' @seealso expand.grid
+#' @param .x,.y,.z Lists or atomic vectors.
+#' @param .l A list of lists or atomic vectors.
+#' @param .wide If \code{TRUE}, returns a list of unique combinations.
+#' If \code{FALSE}, returns a list of the same size as the number of
+#' arguments (2 for \code{cross()}, 3 for \code{cross3()},
+#' \code{length(.l)} for \code{cross_n()}).
+#' @param .filter A predicate function
+#' @return A list or a data frame.
+#' @export
+#' @examples
+#' # We build all combinations of names, greetings and separators from our
+#' # list of data and pass each one to paste()
+#' data <- list(
+#'   id = c("John", "Jane"),
+#'   greeting = c("Hello.", "Bonjour."),
+#'   sep = c("! ", "... ")
+#' )
+#'
+#' data %>% 
+#'   cross_n() %>%
+#'   map(splat(paste))
+#'
+#' # For this purpose, the long format is less pratical and requires a loop
+#' args <- cross_n(data, .wide = FALSE)
+#'
+#' out <- vector("list", length(args[[1]]))
+#' for (i in seq_along(out))
+#'   out[[i]] <- map(args, i) %>% splat(paste)()
+#' out
+#'
+#' # We can filter out unwanted combinations with a predicate function
+#' filter <- function(x, y) x >= y
+#' cross(1:5, 1:5, .filter = filter) %>% str()
+cross <- function(.x, .y, .wide = TRUE, .filter = NULL) {
+  cross_n(list(.x, .y), .wide = .wide, .filter = .filter)
+}
+
+#' @export
+#' @rdname cross
+cross3 <- function(.x, .y, .z, .wide = TRUE, .filter = NULL) {
+  cross_n(list(.x, .y, .z), .wide = .wide, .filter = .filter)
+}
+
+#' @export
+#' @rdname cross
+cross_n <- function(.l, .wide = TRUE, .filter = NULL) {
+  if (!is.null(.filter)) {
+    return(filtered_cross(.l, .wide, .filter))
+  }
+
+  n <- length(.l)
+  remaining <- prod(vapply(.l, length, numeric(1)))
+  rep_factor <- 1
+
+  if (remaining == 0) {
+    return(list())
+  }
+
+  if (.wide) {
+    model <- rep_len(list(NA), n)
+    names(model) <- names(.l)
+    out <- rep_len(list(model), remaining)
+  } else {
+    out <- vector("list", n)
+    names(out) <- names(.l)
+  }
+
+
+  for (i in seq_len(n)) {
+    x <- .l[[i]]
+    x_length <- length(x)
+    remaining <- remaining / x_length
+
+    seq <- rep.int(seq_len(x_length), rep.int(rep_factor, x_length))
+    seq <- rep.int(seq, remaining)
+    rep_factor <- rep_factor * x_length
+
+    if (.wide) {
+      for (j in seq_along(out)) {
+        out[[j]][[i]] <- x[seq][j]
+      }
+    } else {
+      out[[i]] <- x[seq]
+    }
+  }
+
+  out %>% output_hook(.l)
```

Comment:
I think this should only be applied if `!.wide` - it's unlikely to give a useful output if `.wide` is TRUE